File: Linear/Dynamic/Linked List/Singly Linked List/03_Deleting/Given_Position/Deleting.js
import readline from 'readline'

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

function deleteGivenNode(head, pos) {
    let temp = head;

    if (head === null || pos < 1)
        return null;

    // Head is to be deleted
    if (pos === 1) {
        head = head.next;
        temp = null;
        return head;
    }

    // Traverse to the node before the one to be deleted
    let prev = null;
    for (let i = 1; i < pos && temp !== null; i++) {
        prev = temp;
        temp = temp.next;
    }

    if (temp === null)
        return head;

    prev.next = temp.next;

    return head;
}

function printList(head) {
    while (head !== null) {
        process.stdout.write(head.data.toString())
        if (head.next !== null)
            process.stdout.write(" -> ")
        head = head.next
    }
    console.log()
}

let head = new Node(1);
head.next = new Node(2);
head.next.next = new Node(3);
head.next.next.next = new Node(4);

console.log("Linked List before deletion");
printList(head);

rl.question("Enter a position: ", (ans) => {
    const pos = parseInt(ans)

    head = deleteGivenNode(head, pos);

    console.log("Linked List after deletion");
    printList(head);
    
    rl.close()
})

File: Linear/Dynamic/Linked List/Singly Linked List/03_Deleting/Given_Position/Deleting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def deleteGivenNode(head, pos):
    temp = head

    if head is None or pos < 1:
        return None
    
    # Head is to be deleted
    if pos is 1:
        head = head.next
        temp = None
        return head
    
    # Traverse to the node before the one to be deleted
    prev = None
    for i in range(1, pos, 1):
        if temp is None:
            return head
        prev = temp
        temp = temp.next
    
    if temp is None:
        return head
    
    prev.next = temp.next
    return head

def printList(head):
    while head is not None:
        print(head.data, end="")
        if head.next is not None:
            print(" -> ", end="")
        head = head.next
    print()

head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = Node(4)

print("Linked List before deletion\n")
printList(head)

pos = int(input("Enter a position: \n"))
head = deleteGivenNode(head, pos)

print("Linked List after deletion\n")
printList(head)